packages/transactions/src/codecs/transaction-codec.ts

```diff
@@ -10,3 +29,80 @@ export function getNewTransactionEncoder(): VariableSizeEncoder<NewTransaction>
         ['messageBytes', getBytesEncoder()],
     ]);
 }
+
+export function getNewTransactionDecoder(): VariableSizeDecoder<NewTransaction> {
+    return transformDecoder(
+        getStructDecoder([
+            ['signatures', getArrayDecoder(fixDecoderSize(getBytesDecoder(), 64), { size: getShortU16Decoder() })],
+            ['messageBytes', getBytesDecoder()],
+        ]),
+        decodePartiallyDecodedTransaction,
+    );
+}
+
+export function getNewTransactionCodec(): VariableSizeCodec<NewTransaction> {
+    return combineCodec(getNewTransactionEncoder(), getNewTransactionDecoder());
+}
+
+type PartiallyDecodedTransaction = {
+    messageBytes: ReadonlyUint8Array;
+    signatures: ReadonlyUint8Array[];
+};
+
+function decodePartiallyDecodedTransaction(transaction: PartiallyDecodedTransaction): NewTransaction {
+    const { messageBytes, signatures } = transaction;
+
+    /*
+    Relevant message structure is at the start:
+    - transaction version
+    - transaction header (includes number of signer accounts)
+    - static accounts (signers first)
+    */
+
+    let offset = 0;
+
+    // read version
+    offset = getTransactionVersionDecoder().read(messageBytes, offset)[1];
+
+    // read header
+    // first byte of header is numSignerAccounts
+    const numSignersResult = getU8Decoder().read(messageBytes, offset);
+    const numSigners = numSignersResult[0];
+    // there are two more bytes which we don't need to read
+    offset = numSignersResult[1] + 2;
+
+    // read signer addresses (these are first in the static accounts)
+    // compact-u16 array format, read the short-u16 length
+    const addressesLength = getShortU16Decoder().read(messageBytes, offset);
+    offset = addressesLength[1];
+
+    // by decoding this array with a fixed size,
+    // we don't try to read a length prefix and only read as many addresses as we need
+    const [signerAddresses] = getArrayDecoder(getAddressDecoder(), { size: numSigners }).read(messageBytes, offset);
+
+    // signer addresses and signatures must be the same length
+    // we encode an all-zero signature when the signature is missing
+    if (signerAddresses.length !== signatures.length) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__SIGNATURES_LENGTH_NUM_SIGNERS_MISMATCH, {
+            signaturesLength: signatures.length,
+            signerAddresses,
+            signerAddressesLength: signerAddresses.length,
+        });
+    }
```

Comment:
I definitely prefer your comments (though nit: that n is actually probably always 1 yay short u16). But I can't decide if I prefer this code layout to the mutable offset, since offset is the term used in all the codec read functions.I think the naming of the offset being the thing that comes after is a bit confusing. Not sure though!

Reply:
Sounds good.